modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -0,0 +1,544 @@
+      # To decode the ObjectType we need to do another query to CN=Configuration,DC=daforest,DC=com
+      # and look at either schemaIDGUID or rightsGUID fields to see if they match this value.
+      next unless ace_body[:flags] && ace_body[:flags][:ace_object_type_present] == 1
+
+      object_type = ace_body[:object_type]
+
+      if ace_access_mask.match(/CONTROL_ACCESS/) && (object_type == CERTIFICATE_ENROLLMENT_EXTENDED_RIGHT || object_type == CERTIFICATE_AUTOENROLLMENT_EXTENDED_RIGHT)
+        if ace_string.match(/DENIED/)
+          flag_allowed_to_enroll = false
+        elsif ace_string.match(/ALLOWED/)
+          flag_allowed_to_enroll = true
+          allowed_sids << ace_body[:sid].to_s
+        end
+      end
+    end
+
+    [flag_allowed_to_enroll, allowed_sids]
+  end
+
+  def query_ldap_server(raw_filter, attributes, base_prefix: nil)
+    ldap_connect do |ldap|
+      validate_bind_success!(ldap)
+
+      if !@base_dn.blank?
+        vprint_status("Using already discovered base DN: #{@base_dn}")
+      elsif (@base_dn = datastore['BASE_DN'])
+        print_status("User-specified base DN: #{@base_dn}")
+      else
+        print_status('Discovering base DN automatically')
+
+        unless (@base_dn = discover_base_dn(ldap))
+          print_warning("Couldn't discover base DN!")
+        end
+      end
+
+      if @base_dn.blank?
+        fail_with(Failure::BadConfig, 'No base DN was found or specified, cannot continue!')
+      end
+
+      if base_prefix.blank?
+        full_base_dn = @base_dn.to_s
+      else
+        full_base_dn = "#{base_prefix},#{@base_dn}"
+      end
+      begin
+        filter = Net::LDAP::Filter.construct(raw_filter)
+      rescue StandardError => e
+        fail_with(Failure::BadConfig, "Could not compile the filter to find the ESC1 objects! Error was #{e}")
+      end
+
+      returned_entries = ldap.search(base: full_base_dn, filter: filter, attributes: attributes)
+      query_result = ldap.as_json['result']['ldap_result']
+
+      result_code, result_message = check_query_result_code(query_result, filter)
+      case result_code
+      when -1
+        fail_with(Failure::BadConfig, result_message)
+      when 0
+        vprint_good(result_message)
+      when 1
+        fail_with(Failure::NoAccess, result_message)
+      when 2
+        fail_with(Failure::UnexpectedReply, result_message)
+      end
+
+      if returned_entries.blank?
+        print_error("No results found for #{filter}.")
+
+        nil
+      else
+
+        returned_entries
+      end
+    end
+  rescue Rex::ConnectionTimeout
+    fail_with(Failure::Unreachable, "Couldn't reach #{datastore['RHOST']}!")
+  rescue Net::LDAP::Error => e
+    fail_with(Failure::UnexpectedReply, "Could not query #{datastore['RHOST']}! Error was: #{e.message}")
+  end
+
+  def convert_sids_to_human_readable_name(sids_array)
+    output = []
+    for sid in sids_array
+      raw_filter = "(objectSID=#{sid})"
+      attributes = ['dn', 'description']
```

Comment:
Instead of the distinguished name, why not just use the `sAMAccountName`. That's a pretty human readable name, it's more closely aligned with what the rbcd or samr_computer modules would use as well. The actual location included in the DN isn't really super useful in this context AFAIK.

Reply:
The idea with the DN here was that the object may be located in another known computer or forest or even a domain if the server is in charge of multiple domains. So I thought printing this out may help narrow down where the object is.

As for sAMAccountName, I don't know about that since technically speaking a SID doesn't have to have a sAMAccountName associated with it I don't think? I'll do a double check on this though.

Reply:
Fixed with 42b440ada8c5ed95c633cb00b015382fcb11b6c9